 # 1) Create a MLP model with 10 hidden layers using  "data.csv" dataset and report performances using appropriate metrics.

In [4]:
import pandas as pd
dataset = pd.read_csv("data_10.csv")
dataset.info()
dataset.isnull().values.any()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43876 entries, 0 to 43875
Columns: 101 entries, t_0 to malware
dtypes: int64(101)
memory usage: 33.8 MB


False

In [5]:
target_dataframe = dataset.pop('malware')
target_dataframe.head()

0    1
1    1
2    1
3    1
4    1
Name: malware, dtype: int64

In [6]:
from sklearn.preprocessing import StandardScaler
standardScaler = StandardScaler()
standardScaler.fit(dataset)
scaled_dataset = pd.DataFrame(standardScaler.transform(dataset), columns= dataset.columns.values)
scaled_dataset.head()

t_0       t_1       t_2       t_3       t_4       t_5       t_6  \
0 -0.407815  1.040624  0.205229  0.381018  1.214753 -0.246378  0.806037   
1 -0.804648 -0.057878  0.813612  0.278256 -0.217917 -0.870304  0.030828   
2 -1.677680 -1.688988  1.925485 -1.057655  0.737197 -0.870304  1.256741   
3 -0.804648 -0.057878  0.813612  0.278256 -0.217917 -0.870304  0.030828   
4 -0.804648  0.474729 -0.654899  0.748027 -0.990435  1.001474 -0.960719   

        t_7       t_8       t_9  ...      t_90      t_91      t_92      t_93  \
0  1.510980 -0.998545  0.760494  ... -0.521894 -1.065181  1.496827 -0.336061   
1 -1.253474  0.439525 -0.942339  ... -1.237039 -0.937536  0.807549 -0.578494   
2 -1.253474  1.458157 -0.942339  ... -0.446616 -1.141769 -0.740302 -0.497683   
3 -1.253474  0.439525 -0.942339  ...  0.707653  0.683564  1.557290  0.647142   
4  0.625133 -0.384369  0.086846  ... -1.487966  0.696328  1.049401 -1.615571   

       t_94      t_95      t_96      t_97      t_98      t_99  
0  0.090902  0.585610 -1.470810  0.634203 -1.396636 -1.072355  
1 -1.332798  1.746557 -0.217968  0.160702  0.723982 -1.527663  
2 -1.418220 -0.872652 -0.368309 -1.195815 -0.636415 -0.553809  
3  1.955950  0.189190  0.696607  1.837152  0.897365  1.849207  
4  0.631908  1.222715 -0.142797  1.299664 -0.262973  1.318014  

[5 rows x 100 columns]

In [7]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(scaled_dataset, target_dataframe, test_size=0.20, random_state=42)

In [8]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [9]:
model = Sequential()
model.add(Dense(100, activation='relu', input_dim=100))
for _ in range(10):
    model.add(Dense(50, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])
model.fit(X_train, y_train, epochs=10, batch_size=32, validation_split=0.2)
loss, accuracy = model.evaluate(X_test, y_test)
print(f"Loss: {loss}")
print(f"Accuracy: {accuracy}")

Epoch 1/10


2023-09-27 19:18:23.689227: W tensorflow/tsl/platform/profile_utils/cpu_utils.cc:128] Failed to get CPU frequency: 0 Hz


878/878 [==============================] - 1s 984us/step - loss: 0.0819 - accuracy: 0.9792 - val_loss: 0.0506 - val_accuracy: 0.9855
Epoch 2/10
878/878 [==============================] - 1s 846us/step - loss: 0.0472 - accuracy: 0.9844 - val_loss: 0.0512 - val_accuracy: 0.9860
Epoch 3/10
878/878 [==============================] - 1s 855us/step - loss: 0.0361 - accuracy: 0.9877 - val_loss: 0.0508 - val_accuracy: 0.9858
Epoch 4/10
878/878 [==============================] - 1s 872us/step - loss: 0.0262 - accuracy: 0.9908 - val_loss: 0.0500 - val_accuracy: 0.9869
Epoch 5/10
878/878 [==============================] - 1s 878us/step - loss: 0.0209 - accuracy: 0.9932 - val_loss: 0.0504 - val_accuracy: 0.9853
Epoch 6/10
878/878 [==============================] - 1s 842us/step - loss: 0.0182 - accuracy: 0.9939 - val_loss: 0.0557 - val_accuracy: 0.9869
Epoch 7/10
878/878 [==============================] - 1s 832us/step - loss: 0.0188 - accuracy: 0.9944 - val_loss: 0.0610 - val_accuracy: 0.9842
Epo

 # 2) Analyse impact of different activation function with adam solver on the model.


In [10]:
functions = ["relu", "tanh", "sigmoid", "linear", "selu", "elu"]


In [11]:
for function in functions:
    model = Sequential()
    model.add(Dense(100, activation='relu', input_dim=100))
    for _ in range(10):
        model.add(Dense(50, activation='relu'))
    model.add(Dense(1, activation='sigmoid'))

    model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])
    model.fit(X_train, y_train, epochs=10, batch_size=32, validation_split=0.2)
    loss, accuracy = model.evaluate(X_test, y_test)
    print(f"For function {function} we have:")
    print(f"Loss: {loss}")
    print(f"Accuracy: {accuracy}")

Epoch 1/10
878/878 [==============================] - 1s 924us/step - loss: 0.0837 - accuracy: 0.9780 - val_loss: 0.0620 - val_accuracy: 0.9801
Epoch 2/10
878/878 [==============================] - 1s 840us/step - loss: 0.0490 - accuracy: 0.9853 - val_loss: 0.0535 - val_accuracy: 0.9853
Epoch 3/10
878/878 [==============================] - 1s 823us/step - loss: 0.0371 - accuracy: 0.9879 - val_loss: 0.0542 - val_accuracy: 0.9863
Epoch 4/10
878/878 [==============================] - 1s 839us/step - loss: 0.0279 - accuracy: 0.9905 - val_loss: 0.0527 - val_accuracy: 0.9860
Epoch 5/10
878/878 [==============================] - 1s 843us/step - loss: 0.0230 - accuracy: 0.9922 - val_loss: 0.0706 - val_accuracy: 0.9873
Epoch 6/10
878/878 [==============================] - 1s 865us/step - loss: 0.0184 - accuracy: 0.9935 - val_loss: 0.0851 - val_accuracy: 0.9869
Epoch 7/10
878/878 [==============================] - 1s 817us/step - loss: 0.0160 - accuracy: 0.9941 - val_loss: 0.0925 - val_accuracy:

The comparison between the different activation functions was:
ReLU Activation Function
Test Loss: 0.1136
Test Accuracy: 98.72%

Tanh Activation Function
Test Loss: 0.1033
Test Accuracy: 98.61%

Sigmoid Activation Function
Test Loss: 0.1609
Test Accuracy: 98.60%

Linear Activation Function
Test Loss: 0.0712
Test Accuracy: 98.48%

SELU Activation Function
Test Loss: 0.1037
Test Accuracy: 98.59%

ELU Activation Function
Test Loss: 0.1889
Test Accuracy: 98.61%

# 3) Explain your findings and report the best performance.

From all the activations we tried the best result was achieved by ReLU activation function. It had a 98.6% precision and 0.1136 test loss. ReLU is generally good for hidden layers in deep neural networks as to it mitigates the vanishing gradient problem.
